RecoTracker/MkFitCore/src/PropagationMPlex.cc

```diff
@@ -732,66 +769,194 @@ namespace mkfit {
-
-      const float pt = 1.f / ipt;
-
-      float cosahTmp = 0., sinahTmp = 0.;
+    }
+    float pt[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      pt[n] = 1.f / ipt[n];
+    }
+    
       //no trig approx here, phi can be large
-      const float cosP = std::cos(phiin), sinP = std::sin(phiin);
-      const float cosT = std::cos(theta), sinT = std::sin(theta);
-      const float tanT = sinT / cosT;
-      const float icos2T = 1.f / (cosT * cosT);
-      const float pxin = cosP * pt;
-      const float pyin = sinP * pt;
+    float cosP[NN];
+    float sinP[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      cosP[n] = std::cos(phiin[n]);
+    }
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      sinP[n] = std::sin(phiin[n]);
+    }
+    float cosT[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      cosT[n] = std::cos(theta[n]);
+    }
+    float sinT[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      sinT[n] = std::sin(theta[n]);
+    }
+    float tanT[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      tanT[n] = sinT[n] / cosT[n];
+    }
+    float icos2T[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      icos2T[n] = 1.f / (cosT[n] * cosT[n]);
+    }
+    float pxin[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      pxin[n] = cosP[n] * pt[n];
+    }
+    float pyin[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      pyin[n] = sinP[n] * pt[n];
+    }
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
 
       //fixme, make this printout useful for propagation to z
       dprint_np(n,
                 std::endl
-                    << "k=" << std::setprecision(9) << k << " pxin=" << std::setprecision(9) << pxin
-                    << " pyin=" << std::setprecision(9) << pyin << " cosP=" << std::setprecision(9) << cosP
-                    << " sinP=" << std::setprecision(9) << sinP << " pt=" << std::setprecision(9) << pt);
-
-      const float deltaZ = zout - zin;
-      const float alpha = deltaZ * tanT * ipt * kinv;
-
-      if (Config::useTrigApprox) {
-        sincos4(alpha * 0.5f, sinahTmp, cosahTmp);
-      } else {
-        cosahTmp = std::cos(alpha * 0.5f);
-        sinahTmp = std::sin(alpha * 0.5f);
+                    << "k=" << std::setprecision(9) << k[n] << " pxin=" << std::setprecision(9) << pxin[n]
+                    << " pyin=" << std::setprecision(9) << pyin[n] << " cosP=" << std::setprecision(9) << cosP[n]
+                    << " sinP=" << std::setprecision(9) << sinP[n] << " pt=" << std::setprecision(9) << pt[n]);
+    }
+    float deltaZ[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      deltaZ[n] = zout[n] - zin[n];
+    }
+    float alpha[NN];
+#pragma omp simd
+    for (int n = 0; n < NN; ++n) {
+      alpha[n] = deltaZ[n] * tanT[n] * ipt[n] * kinv[n];
+    }
+    float cosahTmp[NN] = {0.};
+    float sinahTmp[NN] = {0.};
+    if (Config::useTrigApprox) {
```

Comment:
Worked fine with gcc 10 or gcc 11

Accidentally used icc because I did not `unset INTEL_LICENSE_FILE` and got this error from it
```
RecoTracker/MkFitCore/src/PropagationMPlex.cc(845): internal error: bad pointer
  #pragma omp simd
```
I guess icc does not support constexpr.

Reply:
According to https://en.cppreference.com/w/cpp/compiler_support#cpp17 icc supports constexpr if, so this is a compiler bug (`internal error` usually means a compiler bug).  Since icc is EOL, it's unlikely to be fixed.

Reply:
If I align #pragma omp simd with the for loop the internal error in icc goes away.
```
--- a/RecoTracker/MkFitCore/src/PropagationMPlex.cc
+++ b/RecoTracker/MkFitCore/src/PropagationMPlex.cc
@@ -822,17 +822,17 @@ namespace mkfit {
 
     float cosahTmp[NN];
     float sinahTmp[NN];
-    if constexpr (Config::useTrigApprox) {
-#pragma omp simd
+    if (Config::useTrigApprox) {
+      #pragma omp simd
       for (int n = 0; n < NN; ++n) {
         sincos4(alpha[n] * 0.5f, sinahTmp[n], cosahTmp[n]);
       }
     } else {
-#pragma omp simd
+      #pragma omp simd
       for (int n = 0; n < NN; ++n) {
         cosahTmp[n] = std::cos(alpha[n] * 0.5f);
       }
-#pragma omp simd
+      #pragma omp simd
       for (int n = 0; n < NN; ++n) {
         sinahTmp[n] = std::sin(alpha[n] * 0.5f);
       }
```